gnovm/pkg/gnolang/realm.go

```diff
@@ -151,21 +151,63 @@ func (rlm *Realm) DidUpdate(po, xo, co Object) {
 	// Updates to .newCreated/.newEscaped /.newDeleted made here. (first gen)
 	// More appends happen during FinalizeRealmTransactions(). (second+ gen)
 	rlm.MarkDirty(po)
+
 	if co != nil {
+		// If this is a self assignment, then the reference count is decremented again below
+		// when xo's reference count is decremented.
 		co.IncRefCount()
-		if co.GetRefCount() > 1 {
-			if co.GetIsEscaped() {
-				// already escaped
-			} else {
-				rlm.MarkNewEscaped(co)
-			}
-		} else if co.GetIsReal() {
+		if co.GetRefCount() > 1 && !co.GetIsEscaped() {
+			rlm.MarkNewEscaped(co)
+		}
+
+		// Marking a value as dirty and increasing its reference count are mutually exclusive operations.
```

Comment:
true for rlm.MarkDirty(po), but for co, the only time we need to mark it is if after IncRefCount() the refcount is 1 (from be temporarily detached to 0). It is already marked as deleted, but it won't be deleted when the time comes if refcount > 0. When refcount becomes 1 its ownerID probably changed. If the refcount ends up being > 1, it will be saved as escaped.

Reply:
reverted